### PR TITLE
User timings event

### PR DESCRIPTION
**Pull request by Calanthe**
The 'user timings' event (http://www.html5rocks.com/en/tutorials/webperformance/usertiming/) is now fired when all of the dfp ads are fully rendered. That event should be visible now on the webpagetest (http://www.webpagetest.org/result/150112_BP_XFK/1/details/).

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/commercial/dfp.js [hunk 1]
```diff
@@ -281,6 +286,18 @@ define([
...
+                if (key === slot) {
+                    slots[key].rendered = true;
+                    return;
+                }
+            });
+
+            if (_.every(slots, 'rendered')) {
+                userTiming.mark('Ads rendered');
+            }
+        },
```
Comment (phamann) [inline static/src/javascripts/projects/common/modules/commercial/dfp.js]: This function could be simplified to:
```js
var rendered = _.some(slots, function(value, key) {
  return (key === slot)
});

if(rendered) { userTiming('gu.ads.rendered'); }
```

**Comment by phamann** [comment]
Love this. Couple of points:
- We have a naming convention for user timing marks, i.e. prepend with `gu.`
- You assume that they will have been rendered when this function is invoked, what if they haven't? I.e. could you sent a timer or interval?

**Comment by phamann** [comment]
:+1: Other than comments.

File: static/src/javascripts/projects/common/modules/commercial/dfp.js [hunk 2]
```diff
@@ -267,7 +269,10 @@ define([
         },
         parseAd = function (event) {
             var size,
-                $slot = $('#' + event.slot.getSlotId().getDomId());
+                slot = event.slot.getSlotId().getDomId(),
```
Comment (ironsidevsquincy) [inline static/src/javascripts/projects/common/modules/commercial/dfp.js]: `slotId` would be a better name

File: static/src/javascripts/projects/common/modules/commercial/dfp.js [hunk 3]
```diff
@@ -281,6 +286,18 @@ define([
                 callbacks[size] && callbacks[size](event, $slot);
             }
         },
+        allAdsRendered = function (slot) {
+            forEach(slots, function (value, key) {
+                if (key === slot) {
+                    slots[key].rendered = true;
+                    return;
+                }
+            });
```
Comment (ironsidevsquincy) [inline static/src/javascripts/projects/common/modules/commercial/dfp.js]: surely just

```slots[slot].rendered = true```

Comment (ironsidevsquincy) [inline static/src/javascripts/projects/common/modules/commercial/dfp.js]: but I wouldn't add a new `rendered` property directly to the slot, as we don't own it. Change slots to something like


```
slots = {
    'dfp-ad--top-above-nav': {
        isRendered: true,
        slot: googletag.slot
    },
   ...
}
```

here's docs for `googletag.slot` btw - https://developers.google.com/doubleclick-gpt/reference#googletag.Slot

**Comment by Calanthe** [comment]
@phamann We have only one userTiming event fired here https://github.com/guardian/frontend/blob/master/static/src/javascripts/bootstraps/app.js#L48 and it doesn't contain any 'gu'. Do I need to change it also?

**Comment by phamann** [comment]
@Calanthe Apologies, I didn't realise that our `utils/userTiming` module does this for you: https://github.com/guardian/frontend/blob/master/static/src/javascripts/projects/common/utils/user-timing.js#L8

File: static/src/javascripts/projects/common/modules/commercial/dfp.js [hunk 4]
```diff
@@ -281,6 +290,18 @@ define([
                 callbacks[size] && callbacks[size](event, $slot);
             }
         },
+        allAdsRendered = function (slotId) {
+            forEach(slots, function (value, key) {
+                if (key === slotId) {
+                    slots[slotId].isRendered = true;
+                    return;
+                }
+            });
```
Comment (ironsidevsquincy) [inline static/src/javascripts/projects/common/modules/commercial/dfp.js]: you're looping, but not using `value` or `key` anywhere, i.e. all of this is this same as

```
slots[slotId].isRendered = true;
```

**Comment by ironsidevsquincy** [comment]
bar comment, :+1:

**Pull request merged**